### ISSUE TITLE
Feature Request: Show all errors and warnings in project for all files, not just opened ones

### ISSUE THREAD

**Issue by kevinjreece** [Microsoft/vscode#13953]
I am using VS Code on a project at work that has hundreds of files and many layers of nested sub-directories. I frequently make changes that break many files, such as changing the call signature of a commonly used method. Since the project is entirely typescript, it is useful to be able to open the "Problems" view and see the errors and warnings my change caused in the files I have open. However, because of the size of the project, I still need to go to my terminal and run our own make commands to see the list of problems. I then need to perform the fun dance between my terminal and VS Code, searching for the right line in the right file before I can fix the problem.

What I need is a way to tell the "Problems"  view to show me all errors and warnings across all open and closed files in my project. This would allow me to see all problems in a single list and quickly click through and fix them.

**Comment by dbaeumer** [comment]
@waderyan this is a feature request for the builder work we discussed with the TS team.

**Comment by waderyan** [comment]
For reference here is a link to the issue - https://github.com/Microsoft/TypeScript/issues/11229

**Comment by plievone** [comment]
Well, I just use a watching build task with `"problemMatcher": "$tsc-watch"` (see https://code.visualstudio.com/Docs/editor/tasks#_processing-task-output-with-problem-matchers) and all errors and warnings are properly shown in the Problems view. It is quite a nice workflow because changes in open files are noticed immediately due to language server, but save triggers incremental compilation (with `tsc --watch`) which still takes a while, but I can control that latency by saving/not saving yet.

**Comment by maxime1992** [comment]
any ETA on this ? There's a lot of issues opened for that feature and not sure if I missed anything or not :smile:

**Comment by waderyan** [comment]
@maxime1992 thanks for the follow up. This is on our radar. No firm timeline yet.

**Comment by johnfn** [comment]
Something I didn't realize was that it's possible to do this with the Tasks infrastructure that VSCode has. All you have to do is put this into your tasks.json:

```
{
  "version": "0.1.0",
  "command": "tsc",
  "isShellCommand": true,
  "args": ["-w", "-p", "."],
  "showOutput": "silent",
  "isBackground": true,
  "problemMatcher": "$tsc-watch"
}
```

Then just run it and you'll get all of the errors across your entire project in the problems view. 

I'm a bit perplexed why this isn't displayed more prominently. It's super useful and cool.

**Comment by WilliamChelman** [comment]
@johnfn Great solution, thanks a lot. I would even add the "--noEmit" argument if the sole purpose of the task is to show the errors
```json
{
  "version": "0.1.0",
  "command": "tsc",
  "isShellCommand": true,
  "args": ["-w", "-p", ".", "--noEmit"],
  "showOutput": "silent",
  "isBackground": true,
  "problemMatcher": "$tsc-watch"
}
```

**Comment by apperside** [comment]
Hi,
I am trying to use this solution but I have a problem: if I try to run from command line tsc, I have the proper output from all files according to tsconfig.json.

On the problem view instead, I see only the problems from the opened files as you can see in the attached video, is that the expected behaviour?

![ezgif-3-4eff27ef25](https://user-images.githubusercontent.com/5955338/27760654-207429a8-5e4d-11e7-91fe-0cf12eea1acd.gif)

**Comment by adius** [comment]
@kevinjreece You can use the Semantic History feature of iTerm (https://www.iterm2.com/documentation-one-page.html) to prevent the "fun dance between my terminal and VS Code". When cmd clicking on the filename + line number it automatically opens the corresponding line in vscode!

**Comment by pzuraq** [comment]
Anyone know if it's possible to open files with errors/in the Problems panel with a keyboard shortcut? Real pain to keep switching between mouse and keyboard, even after running the task.

**Comment by peabnuts123** [comment]
You can open the Problems panel with `Ctrl+Shift+M` (or `Cmd+Shift+M`), it's under Keyboard Shortcuts as "Show Problems" - is that what you mean?

**Comment by pzuraq** [comment]
No, I'd like a shortcut to go to the next file that has an error, whether or not its open.

I was looking into making an extension to do it, but the API is locked down tight, commands don't have access to the MarkerService (tracks problems) in any way that I can see.

**Comment by peabnuts123** [comment]
Oh I see. That could be quite useful. Sounds like the `Problems` area of vscode in-general has a lot of opportunity to improve.

**Comment by cloudkite** [comment]
anyone experiencing this issue https://github.com/Microsoft/vscode/issues/34412?
if the exact same error occurs between subsequent builds it ignores it

**Comment by cr0cK** [comment]
The trick with `tsc` into a debug task is not always possible, for example in a Vue project with custom Webpack loaders to handle .vue files (`tsc` can't parse Vue files out of the box).

But you can find a way to have all errors by watching and compiling your project into a VS Code terminal.

For example, I did this npm script:
`"live-check": "webpack --config ./build/webpack.dev.conf.js -w --display none",`

And I launch it via "npm run live-check" into a VS terminal and I have all errors in real time.

**Comment by apperside** [comment]
even with the solution from @johnfn , I can only see errors from open files :(.
I see the task running in the output tab, but does not show any error

tasks.json:

{
    "version": "0.1.0",
    "command": "tsc",
    "isShellCommand": true,
    "args": ["-w", "-p", ".", "--noEmit"],
    "showOutput": "silent",
    "isBackground": true,
    "problemMatcher": "$tsc-watch"
  }

tsconfig.json

{
  "compilerOptions": {
    /* Basic Options */
    "target": "es2015",                       /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', or 'ESNEXT'. */
    "module": "es2015",                       /* Specify module code generation: 'commonjs', 'amd', 'system', 'umd' or 'es2015'. */
    // "lib": [],                             /* Specify library files to be included in the compilation:  */
    "allowJs": true,                          /* Allow javascript files to be compiled. */
    // "checkJs": true,                       /* Report errors in .js files. */
    "jsx": "react-native",                    /* Specify JSX code generation: 'preserve', 'react-native', or 'react'. */
    // "declaration": true,                   /* Generates corresponding '.d.ts' file. */
    "sourceMap": true,                        /* Generates corresponding '.map' file. */
    // "outFile": "./",                       /* Concatenate and emit output to single file. */
    "outDir": "./build",                      /* Redirect output structure to the directory. */
    "rootDir": "./src",                       /* Specify the root directory of input files. Use to control the output directory structure with --outDir. */
    "removeComments": false,                  /* Do not emit comments to output. */
    // "noEmit": true,                        /* Do not emit outputs. */
    // "importHelpers": true,                 /* Import emit helpers from 'tslib'. */
    // "downlevelIteration": true,            /* Provide full support for iterables in 'for-of', spread, and destructuring when targeting 'ES5' or 'ES3'. */
    // "isolatedModules": true,               /* Transpile each file as a separate module (similar to 'ts.transpileModule'). */

    /* Strict Type-Checking Options */
    "strict": true,                           /* Enable all strict type-checking options. */
    "noImplicitAny": true,                    /* Raise error on expressions and declarations with an implied 'any' type. */
    "strictNullChecks": true,                 /* Enable strict null checks. */
    "noImplicitThis": true,                   /* Raise error on 'this' expressions with an implied 'any' type. */
    "alwaysStrict": true,                     /* Parse in strict mode and emit "use strict" for each source file. */

    /* Additional Checks */
    // "noUnusedLocals": true,                /* Report errors on unused locals. */
    // "noUnusedParameters": true,            /* Report errors on unused parameters. */
    // "noImplicitReturns": true,             /* Report error when not all code paths in function return a value. */
    // "noFallthroughCasesInSwitch": true,    /* Report errors for fallthrough cases in switch statement. */

    /* Module Resolution Options */
    "moduleResolution": "node",               /* Specify module resolution strategy: 'node' (Node.js) or 'classic' (TypeScript pre-1.6). */
    // "baseUrl": "./",                       /* Base directory to resolve non-absolute module names. */
    // "paths": {},                           /* A series of entries which re-map imports to lookup locations relative to the 'baseUrl'. */
    // "rootDirs": [],                        /* List of root folders whose combined content represents the structure of the project at runtime. */
    // "typeRoots": [],                       /* List of folders to include type definitions from. */
    // "types": [],                           /* Type declaration files to be included in compilation. */
    "allowSyntheticDefaultImports": true      /* Allow default imports from modules with no default export. This does not affect code emit, just typechecking. */

    /* Source Map Options */
    // "sourceRoot": "./",                    /* Specify the location where debugger should locate TypeScript files instead of source locations. */
    // "mapRoot": "./",                       /* Specify the location where debugger should locate map files instead of generated locations. */
    // "inlineSourceMap": true,               /* Emit a single file with source maps instead of having a separate file. */
    // "inlineSources": true,                 /* Emit the source alongside the sourcemaps within a single file; requires '--inlineSourceMap' or '--sourceMap' to be set. */

    /* Experimental Options */
    // "experimentalDecorators": true,        /* Enables experimental support for ES7 decorators. */
    // "emitDecoratorMetadata": true,         /* Enables experimental support for emitting type metadata for decorators. */
  },
  "include": [
    "src/**/*",
    "./node_modules/apperside-react-native-dev-kit/**/*"
  ],
  "exclude": [
    "__tests__",
    "index.android.js",
    "index.ios.js",
    "build",
    "local_history",
    "node_modules"
  ]
}


Why can't I see errors from closed files (I am sure that there are errors) ?

**Comment by chandanch** [comment]
👍

**Comment by eladcandroid** [comment]
As @apperside said.. the task only show errors when the file has changed..

**Comment by enko** [comment]
With the new feature that error are displayed in the workspace in the sidebar, for this working out of the box would be greatly appreciated.

**Comment by ekulabuhov** [comment]
Seems like the task format changed a bit in the latest versions of vscode. New format is:

```
		{
			"label": "Monitor TS Errors",
			"command": "./node_modules/.bin/tsc",
			"type": "shell",
			"args": ["--watch", "--project", "."],
			"presentation": {
				"echo": true,
				"reveal": "always",
				"focus": false,
				"panel": "shared"
			},
			"isBackground": true,
			"problemMatcher": "$tsc-watch"
		}
```

**Comment by ekulabuhov** [comment]
Do you start the task each time you open VSCode? Or is there a way to automatically launch it?

**Comment by chandanch** [comment]
I guess the latest version of VS code shows the count of errors per file 

![capture](https://user-images.githubusercontent.com/8663552/36241378-fbfa2340-123b-11e8-9644-3e359ae780e1.JPG)

**Comment by pabloli** [comment]
How can i disable this?

**Comment by jbrodriguez** [comment]
This is what I'm using 

```
		{
			"label": "TSCompileAll",
			"type": "shell",
			"command": "./node_modules/.bin/tsc --watch --noEmit --project .",
			"problemMatcher": ["$tsc-watch"]
		}
```

It's invaluable, I wonder if it could be made default or at least a setting.

Is it possible to autostart this task on a workspace or project ?

**Comment by molinx** [comment]
I use this:
```
{
    "version": "2.0.0",
    "tasks": [
        {
            "label": "tsc watch",
            "type": "shell",
            "command": "./node_modules/.bin/tsc",
            "isBackground": true,
            "args": ["--watch", "--noEmit", "--project", "www"],
            "group": {
                "kind": "build",
                "isDefault": true
            },
            "presentation": {
                "reveal": "never",
                "echo": false,
                "focus": false,
                "panel": "dedicated"
            },
            "problemMatcher": "$tsc-watch"
        }
    ]
}
```
I use this extension to run the task on startup: https://marketplace.visualstudio.com/items?itemName=yukidoi.blade-runner

I still don't know how to keep the errors of a file (in the Problems panel) when closing it.

**Comment by sillero** [comment]
Having an option to keep the problems listed in the panel after you close a file would be very helpful.

**Comment by daarong** [comment]
Does somebody have a workaround for this issue?  For medium to large projects it can be quite cumbersome to open every file to ensure no errors are present.

Is there a way to open all files at once with one command?  That would be an acceptable workaround to me.

**Comment by Darkle** [comment]
@daarong I usually have [eslint-watch](https://github.com/rizowski/eslint-watch) running in the terminal while I'm coding in vscode to check the whole project.

**Comment by pahofmann** [comment]
Does anyone have a workaround for this for java files?  
I want to see all errors of all files in the problems panel at any time, not only the errors of open files.

**Comment by dakt** [comment]
Is this a joke? I have about 2,000 files should I go and open all of them manually to search for lint errors???

**Comment by Domvel** [comment]
Is there a way to trigger ts lint in vs code? I could use e. `tslint` to display it in the console. 
It would be great if I could trigger VS Code to scan all files in src folder for errors (syntax and linter)

**Comment by simeyla** [comment]
About a month ago I had the little red icons next to my files in the workspace viewer. Now they're gone and I have no idea how I did it. Either a new version of VSCode changed the way it works or I installed an extension that broke it. How to get it working again? I've never even heard of tasks.json and the fact that it worked before suggests to me it is now the default behavior.

**Comment by ajayRaghav37** [comment]
Here is how you can see all problems in less than 10 seconds.

You use a little trick.

Open replace all in files (Ctrl + Shift + H).

Replace ; with ;

Hit replace all. That's it. Now, check Problems.

The assumption here is that all files have at least one semicolon. For bigger projects, if you get a warning  asking you to refine your search, just enter something that is common in all files but is not present a lot.

**Comment by bearoutthere** [comment]
Any update on this?  Although Ajay's shortcut work it feels rather hacky.   Seem like this is an obvious and easy win for VSC.  A switch "Show all errors / Show errors for open files"  would be all that's required.

**Comment by mateuscb** [comment]
Another thought, instead of having this automatically happen on file change. I would be totally OK with having a button in the bottom bar or a command that I can explicitly run that would check all files.  Just another thought.

**Comment by neoncom** [comment]
It is correct, I think that the issue has to do with coordinating TS service that checks all opened files and on the other hand checking the files that are closed and are not normally checked automatically, is that correct? So the easy solution is to ensure that problems pane recieves all TS problems from a single source, is that correct?

Seeing how the problem is not solved by the vscode team (?), a proposal:

Should it not be possible to solve this via an addon or via vscode task problem matcher? Because then you could ensure what and when the problems pane shows anyway all by yourself? Then you would just have to somehow disable vscode from showing its stuff? Can a vscode dev tell me if something like this would be workable?

However I still think this basic stuff should be handled by vscode itsself. It's not like TS errors solve themselves when you close files...

**Comment by neoncom** [comment]
> Another thought, instead of having this automatically happen on file change. I would be totally OK with having a button in the bottom bar or a command that I can explicitly run that would check all files.

You can already do that with a tsc task and problem matcher. Then you just rightclick the file paths in the console to jump to the file. Also please don't make it easier for the vscode team than it probably already should be :)

**Comment by JonNeat** [comment]
How can this not be the most critical of all issues to solve first? The IDE is broken until fixed.

**Comment by JonNeat** [comment]
Okay so this is a hacky solution.
Use @plievone solution by making a task use $tsc-watch. Then get a plugin like AutoLaunch to run the task every time the Visual Studio Code is opened. (https://marketplace.visualstudio.com/items?itemName=philfontaine.autolaunch)

**Comment by uglycoyote** [comment]
to expand on @mateuscb's suggestion, maybe something in the context menu of the file explorer would be good where you could right click on a particular folder and do "find problems in folder" would be a good option, as it would allow the user to decide the trade off between scope versus speed.

**Comment by yyynnn** [comment]
Dont do @ajayRaghav37 suggestion, it will hang vscode.
Damn, i will not gonna write ts code ever again in future)

**Comment by matthewvolk** [comment]
Any updates or ideas of when this will be put into the pipeline? Would be great to have this feature out of the box.

**Comment by ajayRaghav37** [comment]
ES-Lint has introduced a new task in VS Code. You have to enable it in the workspace setings.

`"eslint.provideLintTask": true`

Just go to terminal menu and select run task, then choose

> eslint: lint whole folder.

You can also auto-fix most problems by running the following command in the terminal:

`.\node_modules\.bin\eslint.cmd --fix .`

Reference: https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint

While we still await the problem scanner in VS Code, this is a good enough alternative if you use eslint.

**Comment by atkeranov** [comment]
Much needed feature for larger projects indeed. Any updates?

**Comment by hitchcott** [comment]
@ajayRaghav37, it looks like eslint isn't able to catch typescript errors, just linting rules:

https://github.com/typescript-eslint/typescript-eslint/issues/36#issuecomment-478820127

**Comment by dipacs** [comment]
+1

**Comment by shivawu** [comment]
Up this. I also think that it should be possible to disable the builtin TS checking and linting and just use the results from the build task.

**Comment by hasansaghir** [comment]
+1

**Comment by FishOrBear** [comment]
I now run `tsc --noEmit -w` to solve this problem.

But I think it would be better to share a `ts-server` with VSCode.

But why hasn't it been implemented for so long?

**Comment by Domvel** [comment]
@hasansaghir Another guy who does not understand how Github works. Instead of writing a new post with only `+1` you should hit the `+1` (Thumb Up) button on the first post and or other posts.

**Comment by TomasHubelbauer** [comment]
Is anyone here familiar with the TS language server? Does the LSP API lack a way to communicate this to the IDE? Is this a performance concern? Clearly people are passionate about the lack of this feature (I am), so I think there would be interest in community help on getting this past the finish line.

But that is if the problem here is just the lack of manpower on the VS Code team.

If the issue is different (larger impact on the VS Code architecture, a change to LSP required, unclear performance impact, UX considerations), knowing this would help explain why this seemingly straightforward feature takes so much time.

**Comment by martinog2** [comment]
I see various comments in this issue about how the responsibility of this is on the TypeScript team and not the Visual Studio Code team. But this is also a problem with VS Code's built in '@ts-check' functionality. The workarounds mentioned in this post to use a task executing the 'tsc' command don't work for me because I don't have tsc installed (I'm only using @ts-check and jsdoc).

Before, I used to click every file in my project to find the errors. That was tedious, but at least the problems didn't disappear from the side bar once I closed the file. After one of the updates of the past few months, this problem was introduced: #73153.

> How can this not be the most critical of all issues to solve first? The IDE is broken until fixed.

I agree.

**Comment by Bardimaswift** [comment]
When?

**Comment by mattwills8** [comment]
The 'accepted' solution above works, but some of the fields are deprecated.

```js
{
      "label": "Watch TS Errors",
      "type": "shell",
      "command": "tsc",
      "args": ["-w", "-p", ".", "--noEmit"],
      "isBackground": true,
      "problemMatcher": "$tsc-watch"
}
```

Works for me.

Also see https://code.visualstudio.com/docs/editor/tasks for creating tasks.json file if you dont have one yet

**Comment by OliverJAsh** [comment]
Is it possible to disable VS Code's built-in TS server when running `tsc` as a build task like this? The reason I ask is because otherwise we get two sets of errors: those from the built in TS server and those from the task, and sometimes they are out of sync.

**Comment by fantaclaus** [comment]
The IDE is broken and unusable for work on Typescript projects. So I have to stick to the version 1.28.2 and disable updates.

Strange, but it worked somehow before. What's happened? And when will it work again? Already an year passed and there is no sign of change...

**Comment by Porkechebure** [comment]
October 2019 and this is still at large.

3 years for a feature which others IDE offer.... NNNNICCCEEEEEEEEEEEEEE

**Comment by kanlukasz** [comment]
> 
> 3 years for a feature which others IDE offer.... NNNNICCCEEEEEEEEEEEEEE

Remember that VSCode is not an IDE. It's something beetwen Editor and IDE.
Anyway, I hope they do it

**Comment by Porkechebure** [comment]
> > 3 years for a feature which others IDE offer.... NNNNICCCEEEEEEEEEEEEEE
> 
> Remember that VSCode is not an IDE. It's something beetwen Editor and IDE.
> Anyway, I hope they do it

Who cares. Yet, I am not even able to put the damn task up and running to check the ts files.

The provided guide seems like learning a whole new fucking programming language.
We do work and produce, I don't have time to waste learning all the webpack (and similar tools) bandwagon of bullshit to put up a script that checks my ts files on a IDE (I CALL IT IDE ok?) that doesn't even supports the language the same creators of the IDE created correctly....


what a bullshit

**Comment by mjomble** [comment]
> what a bullshit

For a bit of perspective:

Microsoft made some software and gave it away for free.
This software is helpful to millions of people who are graceful for it.
They are not forcing you to use this software.

You are angry at them for giving you something for free because it is not as perfect as it theoretically could be.

**Comment by asfernandes** [comment]
Also, he uses that same "IDE"'s language, and the IDE is open source and accept pull requests...

**Comment by sillero** [comment]
I've been watching this for a long time now, still waiting for a better way to track errors and warnings, and I've been using it since public launch.

That being said, VS Code is in my opinion undoubtedly the best IDE for JS/TS development - heck, it might even be the best MS product, second only to Github. There has definitely been improvements in features and performances, and the sheer amount of extensions makes me very happy. Yes, the docs could be better, but you can say that about every tech documentation out there.

I'm sure the VS Code team is constantly working hard on bringing better tooling and integrations. 

And I believe we can still ask for new features or keep bumping this one up in a polite manner.

**Comment by bscaspar** [comment]
Adding on to the workaround - It looks like there's now a pre-configured task to monitor the output of tsc -w. On a mac, Terminal -> Configure Tasks... -> tsc: watch - tsconfig.json. Then hit ctrl+shift+B and run the task.

**Comment by AlexeySemigradsky** [comment]
Three years later... Any news?

**Comment by genevpd** [comment]
In the middle of large-scale refactoring. I didn't realize that by closing all affected files I will no longer see all the errors :-(

**Comment by maxiqboy** [comment]
3 years and still open :dancer:

**Comment by Porkechebure** [comment]
> 3 years and still open

Yay, random downvotes from idiots.

Anyway, this is an annoying issue come on!!! You realize one has hundreds of files in a serious business project right?
How we are supposed to keep track of them, by mind?

It's one of the basic functions for which people use IDEs.....

And now it seems it's propagating also to visual studio and its intellisense from time to time

**Comment by caseyhoward** [comment]
I find the problems tab useless because of this so I never use it. Instead I run some command in the terminal that actually gives me what I need. For example, if I all compiler errors for the project I run `tsc -w`. I either want to see all the problems in my project or none. I don't care if a file happens to be open or not. It's irrelevant.

**Comment by fantaclaus** [comment]
Well, if it is so important for someone to see errors only for open files in `Problems` tab, let it be.

But could you please create another tab, say `Build Output` just like in Visual Studio.

This tab would show errors and warnings from the compiler as a simple one-line text and a double click on it would open the file and place the caret right on the line with the error.

Also `F4`/`Shift-F4` shortcuts would be great to jump to next/prev error (but this is not so crucial)

Just clear this tab completely and fill it again after each compilation. That would be enough.

Or maybe someone could create an extension for this?

**Comment by ricklove** [comment]
I'm using an "Open Matching Files" extension to open all .ts files. It's a bit performance heavy, but at least it gets the job done.

**Comment by Artur2** [comment]
I think it's because of typescript language server is single-threaded and don't behave like C#'s compiler
Still waiting for this feature, because sometimes I opening file and trying to find type by name using command palete with #, and nothing happen 😞

**Comment by mjbvz** [comment]
For those who wish to live on the edge, the next VS Code insiders build introduces a `"typescript.tsserver.experimental.enableProjectDiagnostics"` setting that enables experimental project wide error reporting

![Feb-06-2020 16-15-20](https://user-images.githubusercontent.com/12821956/73990425-ebe80a80-48fd-11ea-84ec-d7fc288db7ba.gif)

Note that this is not production ready!  It has known perf issues for larger projects and may break your intellisense

**If you experience a problem while using this setting, please open a new issue**

**Comment by 0x0006e** [comment]
any news

**Comment by 0x0006e** [comment]
> For those who wish to live on the edge, the next VS Code insiders build introduces a `"typescript.tsserver.experimental.enableProjectDiagnostics"` setting that enables experimental project wide error reporting
> 
> ![Feb-06-2020 16-15-20](https://user-images.githubusercontent.com/12821956/73990425-ebe80a80-48fd-11ea-84ec-d7fc288db7ba.gif)
> 
> Note that this is not production ready! It has known perf issues for larger projects and may break your intellisense
> 
> **If you experience a problem while using this setting, please open a new issue**

good job~

**Comment by RobbieGM** [comment]
Where might we be able to check on the progress of this feature? I am eagerly awaiting its addition to the standard edition of VS code when it's ready.

**Comment by maxpaj** [comment]
> For those who wish to live on the edge, the next VS Code insiders build introduces a `"typescript.tsserver.experimental.enableProjectDiagnostics"` setting that enables experimental project wide error reporting
> 
> ![Feb-06-2020 16-15-20](https://user-images.githubusercontent.com/12821956/73990425-ebe80a80-48fd-11ea-84ec-d7fc288db7ba.gif)
> 
> Note that this is not production ready! It has known perf issues for larger projects and may break your intellisense
> 
> **If you experience a problem while using this setting, please open a new issue**

Thank you all contributors to Visual Studio Code for your amazing work.

**Comment by spacesuitdiver** [comment]
> For those who wish to live on the edge, the next VS Code insiders build introduces a `"typescript.tsserver.experimental.enableProjectDiagnostics"` setting that enables experimental project wide error reporting

How does this work with other linters?

**Comment by vdh** [comment]
I've had an issue where any time I enable `"typescript.tsserver.experimental.enableProjectDiagnostics"`, the Typescript linter eventually goes down a rabbit hole of traversing `node_modules` and digging up hundreds of errors from distribution files, even when `"exclude"` in `tsconfig.json` contains `"node_modules"`. I can't quite figure out how to debug this.

**Comment by Martaver** [comment]
I also get the same effect as @vdh - my VSCode parses my entire node_modules folder, and runs eslint against each file.

My eslint/editor configuration is:
```
    "editor.codeActionsOnSave": {
        "source.fixAll.eslint": true
    },
    "eslint.validate": [
        "javascript",
        "javascriptreact",
        "typescript",
        "typescriptreact"
    ],
    "[javascript]": {
        "editor.formatOnSave": false
    },
    "[javascriptreact]": {
        "editor.formatOnSave": false
    },
    "[typescript]": {
        "editor.formatOnSave": false
    },
    "[typescriptreact]": {
        "editor.formatOnSave": false
    },
    "eslint.enable": true
```

Seems like first order of business really should be that any file with node_modules in its path should be ignored, at least for on-save actions.

**Comment by blackfan23** [comment]
This is a much appreciated feature. 
I got the same behaviour as @Martaver @vdh .
"Solved" it for now by adding the filters 

`!*node_modules, !*.vscode`

to the terminal view. (only visually different). In case someone is interested in playing with this.

<img width="829" alt="Annotation 2020-04-17 184432" src="https://user-images.githubusercontent.com/12564081/79593574-f2ec7f80-80db-11ea-8f53-0568cffa0e37.png">

**Comment by Martaver** [comment]
@blackfan23 seems a bit like sweeping the dust under a rug :D

**Comment by phellipeandrade** [comment]
workaround: add this 
```
"files.exclude": {
    "node_modules": true
 }
``` 
to .vscode settings.json

**Comment by SupremeTechnopriest** [comment]
@Martaver Agreed.  Something like:

```
"typescript.tsserver.experimental.enableProjectDiagnostics.ignores": [
  "**/node_modules/**"
]
```

would be ideal.  Its kind of shitty to see over 1k problems on a project when there are actually 0.

**Comment by vdh** [comment]
@blackfan23 That's not an option, due to all the excess CPU churn caused by the nosedive into `node_modules`

**Comment by arisferyanto** [comment]
@vdh I set [skipLibCheck](https://www.typescriptlang.org/tsconfig#skipLibCheck) to true in tsconfig and "problems" tab stopped showing errors from node_modules.

However, based on [this stackoverflow thread](https://stackoverflow.com/questions/52311779/usage-of-the-typescript-compiler-argument-skiplibcheck), turning on this option degrades type checking. Appreciate if someone from VS code or typescript team can shed some light on the pros and cons of turning this option on.

**Comment by eladcandroid** [comment]
@mjbvz 
Works great! thanks!
Can we execute "Fix all" for these errors with a one command?

**Comment by dmitryivashkin** [comment]
What about other languages, not Typescript? For example, I have hundreds of PHP files in the project and I can see problems reported by Intellephense extension for individual file, but what I really want is to run diagnostic for entire workspace.

**Comment by vdh** [comment]
Yes this also causes issues where ESLint is doing the same nosedive into `node_modules`("ESLint: Failed to load config" as it trawls the directories in there). There really needs to be a blacklist or something instead of trying to patch this bug out per-linter.

**Comment by patroza** [comment]
Excluding the node_modules would be a no1 fix :)

**Comment by patroza** [comment]
This seems to work a boot, but - your millage may vary, and of course, no warranty:

1. https://www.npmjs.com/package/patch-package

2. : `./patches/typescript+3.9.2.patch`
```
diff --git a/node_modules/typescript/lib/tsserver.js b/node_modules/typescript/lib/tsserver.js
index 2b6f035..ac6c9b4 100644
--- a/node_modules/typescript/lib/tsserver.js
+++ b/node_modules/typescript/lib/tsserver.js
@@ -149353,7 +149353,7 @@ var ts;
                     return;
                 }
                 // No need to analyze lib.d.ts
-                var fileNamesInProject = fileNames.filter(function (value) { return !ts.stringContains(value, "lib.d.ts"); }); // TODO: GH#18217
+                var fileNamesInProject = fileNames.filter(function (value) { return !ts.stringContains(value, "lib.d.ts") && !ts.stringContains(value, "node_modules"); }); // TODO: GH#18217
                 if (fileNamesInProject.length === 0) {
                     return;
                 }
diff --git a/node_modules/typescript/lib/tsserverlibrary.js b/node_modules/typescript/lib/tsserverlibrary.js
index 60e8447..9ed7445 100644
--- a/node_modules/typescript/lib/tsserverlibrary.js
+++ b/node_modules/typescript/lib/tsserverlibrary.js
@@ -149503,7 +149503,7 @@ var ts;
                     return;
                 }
                 // No need to analyze lib.d.ts
-                var fileNamesInProject = fileNames.filter(function (value) { return !ts.stringContains(value, "lib.d.ts"); }); // TODO: GH#18217
+                var fileNamesInProject = fileNames.filter(function (value) { return !ts.stringContains(value, "lib.d.ts") && !ts.stringContains(value, "node_modules"); }); // TODO: GH#18217
                 if (fileNamesInProject.length === 0) {
                     return;
                 }
```

= profit?

**Comment by DanTup** [comment]
@mjbvz I tried enabling this, and I see pretty high CPU usage from the "Code Helper (Renderer)" process compared to when I turn it off. Does it seem likely this is related ("Renderer" in the process name suggests not, but I toggled it a few times). The project is not huge (it's https://github.com/Dart-Code/Dart-Code).

It doesn't last forever, but it spins my MacBook fans up enough that right now that I'd probably keep it off rather than on.

**Comment by vdh** [comment]
Has there been any investigations so far into why `"typescript.tsserver.experimental.enableProjectDiagnostics"` always dives into all the files inside `node_modules`, despite our various failed attempts at setting linter configs to avoid `node_modules`?

It's disappointing, because this potentially-awesome feature is completely killed by the excessive CPU churn caused by it iterating unwanted files & directories… 😢

**Comment by hi2u** [comment]
I've been seeing the same issues @vdh mentioned above with TypeScript errors from `node_modules` appearing in the `PROBLEMS` panel - wasn't sure whether to report a bug, or if it's just some mis-config issue I've caused.

I've tried various `tsconfig.json` settings like:

```js
"exclude": ["node_modules", "./node_modules/*"],
```

I haven't quite figured out if there's a pattern to whether I see the issue or not.  Some things I suspect might be/were happening:

* It seemed to randomly come and go, sometimes restarting vscode helped for a bit, but it would come back later.
* Appeared to affect some projects more than others.  
* I'm not seeing the issue right now, so perhaps a recent update fixed something?  I'm still running the Insiders edition.

Not entirely sure about the points above, could just be me being confused.  Just adding another possibly relevant anecdote.  If there's a few of us still seeing the issue in current versions, may be worth filing a new issue.  Haven't done that yet seeing I'm not quite sure about any of it.

**Comment by enko** [comment]
I had the same problem and my solution in #90430  was to use [typescript-tslint-plugin](https://github.com/Microsoft/typescript-tslint-plugin).

**Comment by vdh** [comment]
@enko TSLint has been [deprecated in favour of ESLint since 2019](https://medium.com/palantir/tslint-in-2019-1a144c2317a9), advocating a plugin for a deprecated linter is not a good solution to this issue.

**Comment by jgoux** [comment]
I had to disable `typescript.tsserver.experimental.enableProjectDiagnostics` because it made 
```
"editor.codeActionsOnSave": {
    "source.fixAll.eslint": true
  }
```
extremely slow (2min to save a **single** file).

I also noticed random TS errors from node_modules, which were appearing / disappearing.

**Comment by samesfahani-tuplehealth** [comment]
There is a working fix actually if you set your `applyTo` properly in tasks.

A full working task that will watch compile and report problems to the Problems pane is shown below:

```json
{
  "version": "2.0.0",
  "tasks": [
    {
      "label": "tsc watch",
      "type": "shell",
      "command": "tsc",
      "isBackground": true,
      "args": ["--build", "--watch"],
      "group": {
        "kind": "build",
        "isDefault": true
      },
      "presentation": {
        "reveal": "never",
        "echo": false,
        "focus": false,
        "panel": "dedicated"
      },
      "problemMatcher": {
        "base": "$tsc-watch",
        "applyTo": "allDocuments"
      }
    }
  ]
}
```

**Comment by vdh** [comment]
@samesfahani-tuplehealth Please edit your comment to either say how this task relates directly to `enableProjectDiagnostics`, or else reword that it is a suggestion as a temporary _workaround_, **not** a fix. The distinction between workarounds and real actual fixes is an important one to make when commenting on issue threads.

**Comment by samesfahani-tuplehealth** [comment]
@vdh Fair enough, edited. Though I don't think this Issue in particular was around `enableProjectDiagnostics` either; the issue at hand was that we cannot view problems across the entire project even for files not currently in view.

**Comment by hi2u** [comment]
Regarding the issue with the unwanted `node_modules` errors, I've created a new issue to specifically discuss that: https://github.com/microsoft/vscode/issues/103539

@vhd / @martaver / @blackfan23 / @SupremeTechnopriest / @patroza / @jgoux / anyone else... If you could please add any more additional clues you might to that issue, it might help in getting this solved hopefully.

**Comment by PinkaminaDianePie** [comment]
The same issue as @jgoux, it doesn't work with `"source.fixAll.eslint": true` - eslint starts to check every file in the project, not only open files, and it rechecks entire project on every single change. VSCode just hangs. No idea why does it modify default ESLint behavior, but that's really a blocker

**Comment by SrBrahma** [comment]
"typescript.tsserver.experimental.enableProjectDiagnostics" really should be fixed and improved. It would make typescript way more powerful than already is.

**Comment by phatmann** [comment]
The solution posted by @samesfahani-tuplehealth works super well. This is crucial for me when I move an imported type from one file to another and need to find everywhere the imports broke. I am not seeing high CPU use, but if so you could just run it when you need it.

**Comment by vdh** [comment]
@phatmann That's a **workaround**, _not_ a solution. Bypassing the feature and using _another entirely different feature_ instead is _not_ a way to fix the performance issues. I hate to be persistent about this but I would rather not have a bunch of noise drawing attention away from fixing this great (but experimental) feature. It could become something incredibly useful if it got the attention it deserves.

**Comment by phatmann** [comment]
@Vdh you are right of course. (And in fact a simpler version of the task, the built in tsk:watch task, works equally well for the workaround). My confusion is this: the workaround is so effective and efficient that I feel no need for the actual feature referenced in this issue. Am I missing something?

**Comment by vdh** [comment]
@phatmann If I wanted to run a Typescript watch task, I could already do that. This isn't *just* about running TS checks. It runs *all* of VSCode's checks on every project file. Typescript, ESLint, CSpell, etc… It's really powerful and useful to know that even if I forgot to open a file, I can find errors across the whole project (without running each linter manually across all files, maybe even having to resort to doing it externally in the terminal).

The only major flaw right now is that it goes overboard with running over *literally* every file, including normally-forbidden folders like `node_modules`.

**Comment by ruimarques** [comment]
That moment when you search how to do something with vscode but what you get is a 4yo open issue... 😢

**Comment by fardolieri** [comment]
```json
"typescript.tsserver.experimental.enableProjectDiagnostics": true
```

https://stackoverflow.com/a/60116499/13241345

**Comment by trusktr** [comment]
How much money and where do I send?

**Comment by dinofx** [comment]
For those having CPU issues or problems with this feature scanning node_modules, make sure you do NOT have a shared or "workspace" `tsconfig.json` somewhere like the root of your repository **unless** it includes:

```JSON
"files": []
```

otherwise, the default is to include all TypeScript files located at or beneath the folder containing the tsconfig, which likely will cause it to analyze too many files.

**Comment by phatmann** [comment]
Great advice @dinofx. In addition, you can use `include` and `exclude`, instead of `files`, per the [docs](https://www.typescriptlang.org/docs/handbook/tsconfig-json.html#examples).
```
      "include": ["src/**/*"],
      "exclude": ["node_modules", "test/**/*"],
```

**Comment by dinofx** [comment]
That's correct, but there's no need to exclude something that doesn't match the "include" globs.

`"files": []` is more succinct for a "workspace" tsconfig.json file (one that just references a bunch of other tsconfig.json files).  It's not well documented, but you need to name the config files for packages **and** the workspace `tsconfig.json`, or a lot of features won't work properly (this features, find references, refactoring, etc).

For a shared config that is extended, it's probably better to name it something like `tsconfig.browser.json`, `tsconfig.node.json`, etc.  That naming prevents anything from ever caring about the `"files"` or `"includes"`, so you can likely omit them.

**Comment by vdh** [comment]
@dinofx The file traversal issue isn't restricted to just Typescript. Additionally, not everyone uses a monorepo.

**Comment by cmawhorter** [comment]
hopefully saving someone else the trouble -- even if you are using typescript adding node_modules to the tsconfig exclude doesn't resolve the issue anyway. no matter what i try node_modules is included in the files scanned.

i didn't try adding it to the list of files totally excluded by vscode (like **/.git) because... i need it. so...

**Comment by phatmann** [comment]
As far as I can tell, all the imported node_module files get scanned even if node_modules is excluded.

**Comment by sKopheK** [comment]
> 
> 
> There is a working workaround actually if you set your `applyTo` properly in tasks.
> 
> A full working task that will watch compile and report problems to the Problems pane **even for files not currently in view** is shown below:
> 
> ```json
> {
>   "version": "2.0.0",
>   "tasks": [
>     {
>       "label": "tsc watch",
>       "type": "shell",
>       "command": "tsc",
>       "isBackground": true,
>       "args": ["--build", "--watch"],
>       "group": {
>         "kind": "build",
>         "isDefault": true
>       },
>       "presentation": {
>         "reveal": "never",
>         "echo": false,
>         "focus": false,
>         "panel": "dedicated"
>       },
>       "problemMatcher": {
>         "base": "$tsc-watch",
>         "applyTo": "allDocuments"
>       }
>     }
>   ]
> }
> ```

thanks for the updated task settings. is it possible to have a similar task for non-ts projects, where i only use type checking provided by vs code on behalf of ts (https://www.typescriptlang.org/docs/handbook/type-checking-javascript-files.html)?

**Comment by larsendaniel** [comment]
I recently inherited a large project that has a lot of misspellings, I am trying to find a way to use something like CSpell (the `streetsidesoftware.code-spell-checker` extension) to find them all in the entire project, since opening every single of the hundreds of files is going to be tedious. Unfortunately I haven't found anything yet except for this post :(

**Comment by mrwensveen** [comment]
> 
> 
> I recently inherited a large project that has a lot of misspellings, I am trying to find a way to use something like CSpell (the `streetsidesoftware.code-spell-checker` extension) to find them all in the entire project, since opening every single of the hundreds of files is going to be tedious. Unfortunately I haven't found anything yet except for this post :(

Maybe an extension like [Open Multiple Files](https://marketplace.visualstudio.com/items?itemName=fabiospampinato.vscode-open-multiple-files) could be useful here as a workaround? It's a bit of a kludge and I don't know how it will perform, but worth a try, maybe.

**Comment by leoblum** [comment]
+1

**Comment by SrBrahma** [comment]
VSCode team really needs to work on it.

**Comment by mrousavy** [comment]
@SrBrahma you can also work on this, VSCode is open source.

**Comment by polkovnikov-ph** [comment]
The reason why this issue is still here is that there is no good way to implement it. It would require loading all the files into language service, with proportionally higher memory usage. JetBrains IDEs don't do this either for the same exact reason.

I suppose it would require modifications to the whole language service infrastructure to make it viable. Nevertheless, I'd love to see this working at least in small projects, and that shouldn't be hard to do to anyone who knows vscode internals well. Unfortunately, that's neither me, nor (apparently) SrBrahma.

**Comment by mrousavy** [comment]
I understand that this is a challenge that requires a lot of work, but SrBrahma phrased his sentence disrespectfully towards people who provide this software FOR FREE and many of those are contributors who work on this in their free time. These people do not owe you anything.

**Comment by SrBrahma** [comment]
> Nevertheless, I'd love to see this working at least in small projects, and that shouldn't be hard to do to anyone who knows vscode internals well. Unfortunately, that's neither me, nor (apparently) SrBrahma.

This. I was kinda lazy to give that answer to the one above. It isn't because it's open source that I have the knowledge and the will to spent months to learn from 0 how VSCode works internally with TS to then implement it. I would like to do it? Yes, but I have many others professional and life priorities before it.

I said "VSCode team" because it's a 5 yo issue and surely it's complicated af to anyone outside the VSCode internal team, who deeply know how it works and have years of experience with it.

**Comment by SrBrahma** [comment]
> I understand that this is a challenge that requires a lot of work, but SrBrahma phrased his sentence disrespectfully towards people who provide this software FOR FREE and many of those are contributors who work on this in their free time. These people do not owe you anything.

Wasn't my intention at all to be disrespectful. Maybe it's my english limitation, a bad choice of words or your interpretation to my comment. Anyway, my explanation to why I said it is above, won't discuss this no more to prevent pollution here.

**Comment by cmawhorter** [comment]
> The reason why this issue is still here is that there is no good way to implement it. It would require loading all the files into language service

Just wanted to clarify -- this feature **does exist** (typescript.tsserver.experimental.enableProjectDiagnostics) right now but is unusable for some envs in its current state (e.g. node/typescript).

The reason it's unusable is there is no way to prevent it from scanning some locations (node_modules) and the result is every issue for every dependency.

**Comment by fermmm** [comment]
Maybe someone with the knowledge can create an extension that opens all the files in the project and then closes them, so we get the same thing that we get when we have all files open but without showing the tabs. If this dirty solution in theory is possible I don't understand why something similar is not possible

**Comment by x4th** [comment]
@fermmm This is a good extension for opening files based on a glob pattern. https://marketplace.visualstudio.com/items?itemName=bcanzanella.openmatchingfiles

**Comment by Porkechebure** [comment]
So, I'm following this discussion since I don't even know when, and all microsoft have been able to do is to let the discussion stem in a big discussion of who finds the best workaround?

**Comment by collinmccarthy** [comment]
@Porkechebure I’m not sure how closely Microsoft is associated with this product, maybe someone else on here can comment on that. To me it’s just an open source IDE with their name on it. As such the more complicated the feature request is the less likely it is to be implemented because the contributors do this in their spare time. So I’d say either we should just accept the current work arounds, come up with new ones, or set aside time to learn the code base and contribute. Personally, I’ll just accept the workarounds since I don’t have time to contribute.

Not trying to be argumentative, just trying to set expectations appropriately here. VSCode is by far the best IDE and I’m guessing most people following this thread feel the same otherwise they’d be using something else. Let’s try to give their team a little slack when things aren’t perfect because nothing is ever perfect. 

Cheers,
-Collin

**Comment by nbaleli-w** [comment]
this worked for me:

```
{
  "version": "2.0.0",
  "tasks": [
    {
      "label": "tsc",
      "type": "shell",
      "command": "./node_modules/typescript/bin/tsc",
      "args": ["--noEmit", "--watch"],
      "problemMatcher": [
        "$tsc-watch"
      ],
      "isBackground": true
    }
  ]
}
```

**Comment by dinofx** [comment]
> this worked for me:

No it didn't.  As soon as you open a file with an error and close it, the error disappears from the problems view.  Reopen the file, and it comes back.

**Comment by AcademyOfF** [comment]
> The reason why this issue is still here is that there is no good way to implement it. It would require loading all the files into language service, with proportionally higher memory usage. JetBrains IDEs don't do this either for the same exact reason.

Honestly, I don't know what would be so terrible about just running tsc in the background and parsing the output. I'm not saying it's a good solution, but that's effectively what I want anyway. Show me the errors tsc shows me in the terminal, but I want to be able to click on them to jump to the problem.

**Comment by joehillen** [comment]
@AcademyOfF You're right. The better solution would be to have a dedicated panel that shows the errors for the whole project with the full error message and some surrounding context as separate blocks similar to the [search panel](https://code.visualstudio.com/updates/v1_42#_moving-views-from-the-sidebar-to-the-panel) that opens the file in a split window when clicked. I wonder if an extension for that already exists.

**Comment by JerryGreen** [comment]
That's a huge topic but I don't see anyone ever mentioned this setting:

```
"typescript.tsserver.experimental.enableProjectDiagnostics": true
```

It makes exactly this: VSCode is searching problems through all the project files, rather than just open ones.

Expect much higher CPU/memory consumption, and also, honestly, this feature is kinda bugged, because sometimes it also shows sources of libraries rather than just project code:

- https://github.com/microsoft/vscode/issues/117732

But in overall, it works.

**Comment by sKopheK** [comment]
does it work when TS compiler is not enabled? i.e. pure JS project

**Comment by AcademyOfF** [comment]
> That's a huge topic but I don't see anyone ever mentioned this setting:
> 
> ```
> "typescript.tsserver.experimental.enableProjectDiagnostics": true
> ```

Honestly, I thought this issue was about that feature, and when this issue was closed, it would mean that feature is no longer experimental, but ready for the general public. 

If this issue isn't about that setting, what is it about? That setting solves the problem, it's just buggy in its current state. Is this a duplicate? Is this issue about alternative ways of solving the problem? What's going to happen to this issue when `typescript.tsserver.experimental.enableProjectDiagnostics` has all its kinks ironed out?

**Comment by JerryGreen** [comment]
@AcademyOfF look when this issue was created: 5 years ago. I guess back then there was no such setting at all. Idk why no one has ever mention this experimental feature till now but... Some day it should have happened, right?

Later issue, about the bug that it is searching through library files sometimes, - it is more specific and it is only 1 year old:

- https://github.com/microsoft/vscode/issues/117732

**Comment by Daynil** [comment]
Update to the solutions by @johnfn and @WilliamChelman due to new syntax for tasks.json, and capability of using npm instead of native shell.

If you have a check-types script in package.json:
```json
{
  "scripts":  {
    "check-types": "tsc -w --noEmit"
  }
}
```

You can add the following task to your tasks.json (note if you don't already have a tasks.json, it goes in the root folder/.vscode/tasks.json): 
```json
{
  "version": "2.0.0",
  "tasks": [
    {
      "label": "check types",
      "detail": "Get all typescript errors in problems pane",
      "script": "check-types",
      "type": "npm",
      "isBackground": true,
      "problemMatcher": "$tsc-watch",
      "presentation": {
        "reveal": "silent"
      }
    }
  ]
}
```

Now you can run this task any time you want project-wide type error checking in the problems pane rather than just in open files.

**Comment by JerryGreen** [comment]
@Daynil yeah, but there are other extensions, like css-linters, markdown-linters, etc. And some people use other languages than JS/TS, too. The general solution is:

```
"typescript.tsserver.experimental.enableProjectDiagnostics": true
```

This enables diagnostics for entire project, for whatever extensions which write errors/warnings/info into "Problems" pane.

**Comment by baptisteArno** [comment]
This setting is making my PC lag a lot.

Is there a way to just run a command that runs a diagnostic only once?

**Comment by robertgornicki-codino** [comment]
> This setting is making my PC lag a lot.
> 
> Is there a way to just run a command that runs a diagnostic only once?

Yeah, just a separate command would be insanely useful.

**Comment by fermmm** [comment]
> This has been buried in this discussion by now, but this task will do exactly what you're looking for.
> Open or create .vscode/tasks.json

Your solution continuously watches for changes, I was confused at first about that. Also you can add the task for all projects:
> Tasks: Open User Tasks
> (Select first option)

Add the task there instead. Then:

> Manage Automatic Tasks in Folder
> Allow Automatic Tasks in Folder

I don't know what happens when you open a non typescript project. Probably throws a silent error and no problem.

**Comment by niebloomj** [comment]
![Screen Shot 2021-11-10 at 10 01 45 AM](https://user-images.githubusercontent.com/5156403/141137326-d796127c-c707-43c4-9d23-13acae7624a7.png)

I have now been using workspace mode for python and this is exactly what I am looking for VS Code to have for the rest of the languages. In my opinion, this should be an IDE specific feature like it is for jetbrains as opposed to being only available for certain languages. As someone who uses many languages in the same project, its very unfortunate that I can only get a birds eye view of my python code. Thanks!

**Comment by Porkechebure** [comment]
So, it's years I'm following this thread and many thing have been said. I peeked on in a while and I saw a lot of heat here.
Is there a real solution to this or the only viable hacks are injecting some script and configs around?

**Comment by fermmm** [comment]
> So, it's years I'm following this thread and many thing have been said. I peeked on in a while and I saw a lot of heat here. Is there a real solution to this or the only viable hacks are injecting some script and configs around?

There is no solution without issues involved. I remember reading a comment from a VS Code dev that said they don't find a good way of implementing that, sounded like they don't have the programming skills to code this feature.

**Comment by enekonieto** [comment]
It is quite satisfying to see `0 errors` somewhere in the screen, assuring you have no stupid format/syntax errors in you project.

Come on Microsoft, it is your language and it is your code editor!

**Comment by niebloomj** [comment]
I have finally hit the point where I want this feature badly enough that I am willing to code it myself. Does anyone have any clues as to how we can implement this? Any pointers for a first time contributor to the project? This seems like such a fundamental feature that I can't imagine VS Code not having it. Anyone want to partner up?

**Comment by jerrygreen** [comment]
@niebloomj https://github.com/microsoft/vscode/issues/13953#issuecomment-886008557

**Comment by niebloomj** [comment]
> @niebloomj [#13953 (comment)](https://github.com/microsoft/vscode/issues/13953#issuecomment-886008557)

This is only for the typescript server, not a native vs code solution for linting whole projects like jetbrains has.

**Comment by jerrygreen** [comment]
@niebloomj by some reason it falls under "typescript.tsserver" category in settings, probably because it’s implemented by TS team and they mindlessly put it there like the rest things they do, but in practice it’s not TS-specific setting at all, it runs my other plugins like markdown linters, etc. - everything that writes into Problems pane.

**Comment by niebloomj** [comment]
@jerrygreen wow. Okay, I will definitely take a look at that then. In that case, we can close out this 5 year old feature request right? I'll test it out, but am I missing something? Does it work with extensions that write to the problem space?

**Comment by fermmm** [comment]
> @jerrygreen wow. Okay, I will definitely take a look at that then. In that case, we can close out this 5 year old feature request right? I'll test it out, but am I missing something? Does it work with extensions that write to the problem space?

I'ts full of issues, that is why this request is still here.

**Comment by jerrygreen** [comment]
It works with extensions. One crucial issue with it, - it sometimes shows errors not just from project code, but from node_modules (or even VSCode source itself sometimes, it seems):

- https://github.com/microsoft/vscode/issues/117732

Quite recently it was closed in favor of one more specific issue:

- https://github.com/microsoft/TypeScript/issues/36664

It is another long-running issue, created by original creator of the feature itself. The issue is almost 2 y.o. now, - it has some new posts a couple weeks ago though, but I don’t know what takes so long to resolve that. It seems it’s just more complex to fix some config or something, and somehow it’s not a priority for VSCode team, that’s why.

Our issue here is even more old, 5 whole years now, we can send it to school soon. Since it’s about feature existence (feature request), - it is unrelated now because feature exists. Related issues with the feature are formulated and are on github, they just take too long to resolve by some reason. I would close our issue here if I could, in favor of more specific existing ones.

The best thing you can do is probably go and nuke that opened specific issue with thumbs ups, lots and lots of thumbs ups. Please don’t flood messages in that issue though, that would be uncivilized, - people are working there, whilst all people doing here is just «wanting». For the rest who actually want to try to resolve the issue themselves, - it has the details (rather than our issue here which describes nothing other than «we want this»).

**Comment by niebloomj** [comment]
@jerrygreen I tested it out and it doesn't seem to work at all... I open a file and the problems pane fills up with issues. I close a file and all those problems go away... Tried with python, go, java, typescript, etc.

**Comment by jerrygreen** [comment]
@niebloomj do you open a folder with VSCode, or just a single file? You can open a whole folder which will be considered a project, and then it works. Or you can open a single file in which case VSCode won’t consider it part of your project, and work like normal (showing issues only when file is open).

**Comment by niebloomj** [comment]
I open a directory that has multiple projects in it. I do not use vs code in single file mode. Thank you. :)

**Comment by mjbvz** [comment]
> @niebloomj by some reason it falls under "typescript.tsserver" category in settings, probably because it’s implemented by TS team and they mindlessly put it there along with their other settings, but in practice this one is not TS-specific setting at all, it runs my other plugins like markdown-linter, css-linter, etc. - everything that writes into Problems pane. [#13953 (comment)](https://github.com/microsoft/vscode/issues/13953#issuecomment-886237859)

This is incorrect: that setting is only for the typescript server and does not effect any other extensions/languages

Reporting errors in all workspace files must be implemented on a per-language basis (and it already can be implemented if a language/extension supports it). I've updated the issue title to make it clear that this issue tracks JavaScript and TypeScript support for reporting errors in unopened project files

**Comment by niebloomj** [comment]
Since many of us thought this issue was tracking errors and warnings for all languages and not just ts/js, can we get a duplicate of this issue for that feature? If we already have a setting that does it for typescript, can we just revert the title to the way it was and this issue can track the feature request of making it work across languages like it does with jetbrains IDE?

**Comment by mjbvz** [comment]
> can we just revert the title to the way it was and this issue can track the feature request of making it work across languages like it does with jetbrains IDE?

That cannot be implemented by VS Code, it must be implemented on a per language/extension basis. The APIs to implement this features already exist, but not every extension can implement it in an efficient way (and the expected behavior also varies between languages) 

If you have want to see this feature in a given language, file a separate issue against that language's extension. We're keeping this one open to track JS/TS support for this since this issue has been focused on JS/TS from the start

**Comment by str** [comment]
I was also following this issue because I also understood this was a new feature request for all languages.  I understand that this "cannot be implemented by VS Code", that's why I understood it was a new feature request.  I never read anywhere that said this feature was only for JS/Typescript.  I now realized that I waited years in the wrong issue.

I think the tightly coupled JS/Typescript confuses many people.  One does not know what requests are for the global IDE, and what requests are just for JS/Typescript.  

> @mjbvz The APIs to implement this features already exist

Could you please share a link to that API documentation for anyone that would like to start a new issue in the respective language plugin?

Thank you.

**Comment by DanTup** [comment]
@str there are not specific APIs language extensions need to handle diagnostics for unopened files, they're just the same as for open files. Errors are currently pushed to VS Code by a language extension (or LSP server).

This is done using the `set` method on [`DiagnosticCollection`](https://code.visualstudio.com/api/references/vscode-api#DiagnosticCollection) for languages not using LSP, and for LSP servers, sending `textDocument/publishDiagnostics` notifications for all files.

**Comment by niebloomj** [comment]
> @str there are not specific APIs language extensions need to use to provide diagnostics for unopened files, they're just the same as for open files. Errors are currently pushed to VS Code by a language extension using the `set` method on [`DiagnosticCollection`](https://code.visualstudio.com/api/references/vscode-api#DiagnosticCollection) or by sending `textDocument/publishDiagnostics` notifications for all files if using an LSP server.

If they are the same APIs for closed or open files, then how does a language server implement this feature? I assume VSCode itself does the removal of the issues in the problem tab when closing an opened file.

**Comment by DanTup** [comment]
> If they are the same APIs for closed or open files, then how does a language server implement this feature? I assume VSCode itself does the removal of the issues in the problem tab when closing an opened file.

That's not correct - VS Code does not clear the errors when a file is closed, it is all done by the extension. There are existing extensions (including Dart - which I work on) that work the way being requested here without any problems. You open a project, it shows all errors from the whole project, and as you open/close files they do not disappear.

I would expect if you file an issue against a languages extension/LSP Server requesting this, they will understand how to implement it (although whether they can easily do so may depend on tools outside of the extension itself that they are using to provide the diagnostics).

**Comment by niebloomj** [comment]
@DanTup Thanks!

Also,

I just wanted to bring this comment out to people's attention.

>  I don't think the tsserver provides an API for compiling the whole project. So before anything could be done here, the typescript project would have to add that first. Though I imagine that would have to run in a separate process so that it doesn't block all the existing functionality as it would likely require quite a bit of time to compile some projects.

https://github.com/typescript-language-server/typescript-language-server/issues/358#issuecomment-1004867435

**Comment by mindplay-dk** [comment]
Couldn't this potentially be fixed from VS Code by merely *telling* the language server that all the files have been opened?

Do we need to wait for TS to fix this? It doesn't look like it's a priority.

Refactoring without this feature is all trial-and-error - it feels *very* outdated. 😐

**Comment by ianwremmel** [comment]
For what it's worth, my solution to this has been:

```bash
tsc | awk -F '(' '/error TS/ {print $1}' | sort | uniq | xargs -n 1 code
```

it effectively runs the linter at least twice, and open files at a time via `code` is not exactly fast, but i've found this is the most effective way to find all of the type errors and view them in vscode.

**Comment by bartkoedijk** [comment]
I am working on a React Native project and I couldn't see any errors besides the current opened file.
Now I can see all errors in the project by doing the following:

1. create tsconfig.json
2. hit **Control + Shift + P**
3. select the following; Tasks: Configure Tasks
4. scroll down to: tsc: watch - tsconfig.json
5. hit **Control + Shift + P**
6. Tasks: Run tasks 
7. select your created task "tsc: watch - tsconfig.json"

My tsconfig (you can also use the one that React Native has in their documentation)
```
{
  "compilerOptions": {
    "experimentalDecorators": true,
    "allowJs": true,
    "allowSyntheticDefaultImports": true,
    "esModuleInterop": true,
    "isolatedModules": true,
    "jsx": "react",
    "lib": ["es6"],
    "moduleResolution": "node",
    "noEmit": true,
    "strict": true,
    "target": "esnext",
    "useDefineForClassFields": true
  },
  "exclude": ["node_modules", "babel.config.js", "metro.config.js", "jest.config.js"]
}
```
I hope that I've helped somebody:)

**Comment by Clonkex** [comment]
For a while my TS project was small enough that I just kept all the files open at once. Later it became big enough to warrant only keeping relevant files open. I made a small change, saw that there were no errors reported and switched over to cmd to run the program. I was quite surprised to see tsc report errors.

Is there some point when we can expect to see project-level TS error reporting in VSC? I find it quite astonishing and more than a little frustrating that this isn't already a thing. It seems to me to be a fundamental feature of any development environment. The whole reason I use Typescript is to avoid programming in the dark ages of guesswork and trial-and-error. No one should have to suffer the tedium of switching to another program to find what methods are available or check for errors.

VSC can already check inter-file dependencies and errors, so what's the difference between checking the whole project and only checking the open files? Experience tells me that things are never as simple as they seem but it _does_ seem simple. When I open a project, check all files (whether they're open or not). When some time has past since I made a change to a file, check that file and all dependent files (whether they're open or not).

**Comment by djmisterjon** [comment]
do vscode have a way to open and close file ? 
we can maybe do a dirty script like 
```js
const files = ['a.ts, b.ts, c.ts' , '....ts'] 
files .forEach(path=> {
	const tab = vscode.open(path)
	tab.focus()
        tab.close()
});
```

**Comment by djmisterjon** [comment]
> Later issue, about the bug that it is searching through library files sometimes, - it is more specific and it is only 1 year old:

working fine after test 20 min ! thanks ! 
and not so cpu killer!
![image](https://user-images.githubusercontent.com/24865815/163680760-fa17f60c-46da-4523-8f58-02c3e6b3c4d4.png)

it show only my project file
```json
  "include": [
    "src",
    "global.d.ts",
  ],
  "exclude": [
    "node_modules",
    "plugins",
    "public"
  ]
```

it the solution a wanted, your post should pinned !

**Comment by mindplay-dk** [comment]
> Couldn't this potentially be fixed from VS Code by merely _telling_ the language server that all the files have been opened?

How about a generic *option* to do this?

Like a checkbox in the "Problems" tab: ✅ **Check entire project**.

This feature wouldn't need to be specific to TS either - if other language services benefit from VS code pretending to a language service that everything is open, that might work for other language services too?

Presumably, language services are *designed* to be asked to open specific files? I'm guessing, but if the language service protocol offers the ability to scan/open specific files in the first place, presumably all language service implementations have this limitation, and would benefit from this feature?

**Comment by ArtemFrontendACTUM** [comment]
Just use WebStorm

**Comment by lipe-dev** [comment]
> Just use WebStorm

I would, really, if it stopped breaking everything else in my setup.

**Comment by ayush-seth** [comment]
Adding `"typescript.tsserver.experimental.enableProjectDiagnostics": true,` to `settings.json` worked for me.

**Comment by douglasg14b** [comment]
This works, but it just shows errors for everything under `node_modules` which is.... very much unwanted.

**Comment by rafsaf** [comment]
> Adding this to `settings.json` worked for me:
> 
> ```
> "typescript.tsserver.experimental.enableProjectDiagnostics": true,
> ```

For anybody still looking here, I can confirm this solution works like a harm.

It's also mentioned here: https://stackoverflow.com/questions/55201424/how-to-get-vscode-to-show-typescript-errors-for-files-not-open-in-the-editor#60116499

**Comment by nathanpovo** [comment]
One problem with the `typescript.tsserver.experimental.enableProjectDiagnostics` setting is that, whenever you re-install you npm packages using `npm ci`, you suddenly get a bunch of errors.

To fix that issue I always have to turn off the setting and turn it back on (change it from `true` to `false` and back to `true`) for the "JS/TS language features" to be re-initiliased.

**Comment by marinaglancy** [comment]
If anybody finds it useful, I created a VSCode extension "Bulk Problem Diagnostics"
https://marketplace.visualstudio.com/items?itemName=MarinaGlancy.bulk-problem-diagnostics

It allows to open all files in a folder and all subfolders in order to review problems. Splits large number of files in batches to prevent unloading.

It has a wide range of settings controlling how many files to open, what file types, what to exclude, how long to wait before opening files.

**Comment by ad-on-is** [comment]
tsconfig.json

```
{
    "compilerOptions": {
        "skipLibCheck": true,
    }
}
```

to disable errors/problems within node_modules

**Comment by Beyim76** [comment]
![‎Umut ile WhatsApp Sohbeti adlı dosyanın kopyası (1).docx](https://github.com/microsoft/vscode/files/12174401/Umut.ile.WhatsApp.Sohbeti.adli.dosyanin.kopyasi.1.docx)

**Comment by Dejee98** [comment]
Whenever I rebuild solution, VS shows me all Errors and Warnings, but I understand that this is probably impractical for very large projects. I'm also a beginner so, sorry if what I said is not the proper way to do things.

**Comment by adamerose** [comment]
Anyone have advice for getting `enableProjectDiagnostics` working? I have it in my settings and still only see errors on files if they're open. 

- Started a fresh project with `npm create vite@latest typescript-sandbox -- --template react-ts` 
- Added a component with an error
- Disabled all extensions
- Commented out all settings aside from `"typescript.tsserver.experimental.enableProjectDiagnostics": true`

**Comment by ChristopherHaws** [comment]
You will always need to open one file to boot up the language service. From my experience, that setting will show you all errors for the ts projects which currently have open files, not all ts projects in the workspace.

**Comment by ADTC** [comment]
@marinaglancy thank you! You're an angel! I like how it opens in a hidden way. I was afraid it's going to open and close a bunch of tabs quickly (visually on my screen), but it just opens the files internally and checks them, then only opens a tab for them if a problem was found. Awesome!

Now, is there a way to automate it, so that it runs either periodically or runs when saving? But maybe that's not a good idea because it takes a while to check all files. My small project itself has 125 files already, imagine one with 3000 files. If only there was a way to go by references.

**Comment by Porkechebure** [comment]
I like that this is actively open and discussed after years.... LMAO
Js and it's ecosystem are some nasty dumpster fire

**Comment by Ploppy3** [comment]
> I like that this is actively open and discussed after years.... LMAO Js and it's ecosystem are some nasty dumpster fire

Because it's not about fixing the existing issue of this experimental feature for TS or JS, it is about implementing a proper solution in the language server API for any language.

Note: Thank you for the spam. Many people have subscrIbed to this issue.

**Comment by marinaglancy** [comment]
It would be a huge performance hit if all enabled extensions constantly analyse ALL files in the workspace for problems. Currently they usually only do it only when a file is open by listening to the "file open" and also "file edited" events

Maybe VSCode could introduce a new event "file waiting to be analysed" and advise all extensions to listen to it (and trigger it as configured)? But it would require changes to absolutely all extensions in the marketplace that report diagnostics. 

@ADTC your question is specific to the "[Bulk problem diagnostics](https://marketplace.visualstudio.com/items?itemName=MarinaGlancy.bulk-problem-diagnostics)" extension. Let's not pollute the discussion here, I created an issue in my own [repo](https://github.com/marinaglancy/vscode-bulk-problem-diagnostics/issues/2) where we can discuss it

**Comment by M7ilan** [comment]
Facing the same issue, but I found a temporary fix. Enable `typescript.tsserver.experimental.enableProjectDiagnostics` in VSCode settings. But, you need to be in a TS or JS file and restart the TS Server using `>TypeScript: Restart TS Server` from the command palette.

**Comment by djmisterjon** [comment]
> I found a temporary fix. Enable `typescript.tsserver.experimental.enableProjectDiagnostics` in VSCode settings. But, you need to be in a TS or JS file and restart the TS Server using `>TypeScript: Restart TS Server` from the command palette.

https://github.com/microsoft/vscode/issues/13953#issuecomment-1100688744

**Comment by M7ilan** [comment]
> > I found a temporary fix. Enable `typescript.tsserver.experimental.enableProjectDiagnostics` in VSCode settings. But, you need to be in a TS or JS file and restart the TS Server using `>TypeScript: Restart TS Server` from the command palette.
> 
> [#13953 (comment)](https://github.com/microsoft/vscode/issues/13953#issuecomment-1100688744)

I still don't have it. Also, my friend did the same thing I did, and it's working fine for him.

**Comment by jthoward64** [comment]
> This works, but it just shows errors for everything under `node_modules` which is.... very much unwanted.
> 
> My tsconfig has:
> 
> ```json
>   "include": ["src"],
>   "exclude": ["node_modules"]
> ```
> 
> Which I would have assumed means it would exclude node_modules...?
> 
> Even if filtering via `!**/node_modules**` this still causes it to indefinitely hog resources, and the list of problems seems to forever grow and shrink without end. It also seems to kill pre-save tasks like lint-fix, causing files to not actually save...

Anyone know a solution for this? `skipLibCheck` doesn't work in my case because a dumb library I am using ships ts files with their package for react-native (which I'm using) and causes tsc to check those files which immediately gives me a hundred errors. (yeah there are other issues with the library, we're trying to migrate away from it)

**Comment by interpolation-0** [comment]
I have the opposite issue with `typescript.tsserver.experimental.enableProjectDiagnostics`.

Files are incorrectly detected to have errors pertaining to an unresolved module, which is actually in the paths option of my `tsconfig.json`.

The errors only go away while the file with the error is currently active in the editor, not just opened.

There are 187 of these files and over 600 problems, so I'd like to keep them closed :)

**Comment by cyanff** [comment]
> I have the opposite issue with `typescript.tsserver.experimental.enableProjectDiagnostics`.
> 
> Files are incorrectly detected to have errors pertaining to an unresolved module, which is actually in the paths option of my `tsconfig.json`.
> 
> The errors only go away while the file with the error is currently active in the editor, not just opened.
> 
> There are 187 of these files and over 600 problems, so I'd like to keep them closed :)

seconding this issue, anybody stumbled upon a solution?

**Comment by j1mmie** [comment]
> > I have the opposite issue with `typescript.tsserver.experimental.enableProjectDiagnostics`.
> > Files are incorrectly detected to have errors pertaining to an unresolved module, which is actually in the paths option of my `tsconfig.json`.
> > The errors only go away while the file with the error is currently active in the editor, not just opened.
> > There are 187 of these files and over 600 problems, so I'd like to keep them closed :)
> 
> seconding this issue, anybody stumbled upon a solution?

Also seeing this. Took me a while but I've finally determined the issue is related to `enableProjectDiagnostics`. It doesn't happen when `enableProjectDiagnostics` is false.

One possible solution is to enable the (also experimental) [VSCode core's file watcher](https://code.visualstudio.com/updates/v1_89#_file-watching-handled-by-vs-code-core)?

```
"typescript.tsserver.experimental.useVsCodeWatcher": true,
```

I haven't tested it thoroughly though, YMMV.

**Comment by latel** [comment]
https://github.com/user-attachments/assets/011cbd32-d17d-485c-9372-8a61b695529d

The problems numbers keeps changing like crazy 😢

**Comment by dkattan** [comment]
> Kapture.2025-03-19.at.10.06.48.mp4 
> The problems numbers keeps changing like crazy 😢

@latel I was having this same issue. I am also working on a Vue project and noticed that the cycling starts shortly after the Vue diagnostics show up. This appears to be a known issue. 

https://stackoverflow.com/questions/78126814/vue-volar-extension-in-visual-studio-code-keeps-crashing-the-js-ts-language-s
https://www.reddit.com/r/windsurf/comments/1klj8xu/vue_official_extension_causes_ts_language_service/
https://github.com/vuejs/language-tools/issues/4519

The Windsurf guys claim it is fixed in a newer version of the Vue - Official VS Code extension but evidently it hasn't been published yet to the VS Code Marketplace. 

I was able to get it to stop cycling by downgrading the extension with `code --install-extension vue.volar@1.8.27` however it no longer shows workspace-wide errors which is the entire point of this issue. 

I found the allegedly fixed version on Open VSX and installed it via VSIX but the same problem occurred both in 3.0.0-beta.2 and 3.0.0-beta.3

https://open-vsx.org/extension/Vue/volar/3.0.0-beta.3

Edit: I thought it stopped when I opened just the frontend folder instead of my code-workspace but it started again.

**Comment by johnsoncodehk** [comment]
This issue was quite hard to reproduce for us, and the TS team will be better able to investigate this kind of issue than us. I suggest you enable TS logs (setting `"typescript.tsserver.log": "verbose"`) and report it in the TypeScript repo.

**Comment by jaques-sam-tlv** [comment]
Why is this solution only for typescript/javascript? VScode must have a general setting for this!

**Comment by RangerCoder99** [comment]
How is development so focused on AI, and not this massively important feature...

**Comment by fbf-prog64** [comment]
For those who are still looking for a solution, and if the experimental feature mentioned in https://github.com/microsoft/vscode/issues/13953#issuecomment-583179315 is helpful, I just wanted to mention that the setting has been renamed to:

```json
{
    /*...*/
    "js/ts.tsserver.experimental.enableProjectDiagnostics": true
    /*...*/
}
```

The other name still works, but it has been deprecated.

**Comment by frederikhors** [comment]
@fbf-prog64 does this work good?

**Comment by fbf-prog64** [comment]
> [@fbf-prog64](https://github.com/fbf-prog64) does this work good?

@frederikhors, In my particular case, it has been helpful. These are my observations:

* It does not check `node_modules` in my project.
* If you close a file, their items disappear from the "Problems" panel.
* You can run the check again if you keep the `settings.json` open in a tab, then set `false` to the option, and then `true` (disable and reenable).
* You must have one file open to trigger the check.

Most of these things have been mentioned over the years here, so I was already aware of them when I tried this workaround. 👍
